lab_1_keywords_tfidf/main.py

```diff
@@ -3,103 +3,157 @@
     """
     Composes a frequency dictionary from the token sequence
-
     Parameters:
     tokens (List[str]): Token sequence to count frequencies for
-
     Returns:
     Dict: {token: number of occurrences in the token sequence} dictionary
-
     In case of corrupt input arguments, None is returned
     """
-    pass
+    if not check_input_type(tokens, list, str, False):
+        return None
+    tokens_freqs = {}
+    for i in tokens:
+        tokens_freqs[i] = tokens.count(i)
+    return tokens_freqs
 
 
 def get_top_n(frequencies: dict[str, Union[int, float]], top: int) -> Optional[list[str]]:
     """
     Extracts a certain number of most frequent tokens
-
     Parameters:
     frequencies (Dict): A dictionary with tokens and
     its corresponding frequency values
     top (int): Number of token to extract
-
     Returns:
     List[str]: Sequence of specified length
     consisting of tokens with the largest frequency
-
     In case of corrupt input arguments, None is returned
     """
-    pass
+    if not check_input_type(frequencies, dict, str, Union[int, float], False) \
+            or not check_num(top):
+        return None
+    sorted_tokens_freqs = {k: v for k, v in sorted(frequencies.items(), key=lambda item: item[1], reverse=True)}
+    if len(frequencies) > top:
+        top_freqs_words = list(sorted_tokens_freqs.keys())[:top]
+    else:
+        top_freqs_words = list(sorted_tokens_freqs.keys())
+    return top_freqs_words
 
 
 def calculate_tf(frequencies: dict[str, int]) -> Optional[dict[str, float]]:
     """
     Calculates Term Frequency score for each word in a token sequence
     based on the raw frequency
-
     Parameters:
     frequencies (Dict): Raw number of occurrences for each of the tokens
-
     Returns:
     dict: A dictionary with tokens and corresponding term frequency score
-
     In case of corrupt input arguments, None is returned
     """
-    pass
+    if not check_input_type(frequencies, dict, str, int, False):
+        return None
+    term_freq = {}
+    for k, v in frequencies.items():
+        new_v = v / sum(frequencies.values())
+        term_freq[k] = new_v
+    return term_freq
 
 
 def calculate_tfidf(term_freq: dict[str, float], idf: dict[str, float]) -> Optional[dict[str, float]]:
     """
     Calculates TF-IDF score for each of the tokens
     based on its TF and IDF scores
-
     Parameters:
     term_freq (Dict): A dictionary with tokens and its corresponding TF values
     idf (Dict): A dictionary with tokens and its corresponding IDF values
-
     Returns:
     Dict: A dictionary with tokens and its corresponding TF-IDF values
-
     In case of corrupt input arguments, None is returned
     """
-    pass
+    if not check_input_type(term_freq, dict, str, float, False) \
+            or not check_input_type(idf, dict, str, float, False):
+        return None
+    for k in term_freq:
+        if k in idf:
+            for k in idf.keys():
```

Comment:
why do we need to iterate over idf keys while iterating over term_freq keys? 
could we do all that while iterating over just term_freq keys? 

also, using the same variable in nested cycles is not a good idea as you redefine it as soon as you enter the inner cycle